pennylane/optimize/rotosolve.py

```diff
@@ -132,340 +212,410 @@ class RotosolveOptimizer:
+    Minimization substeps: [-1. -1. -1. -1. -1. -1. -1.]
+    Cost before step: -0.9999999999999996
+    Minimization substeps: [-1. -1. -1. -1. -1. -1. -1.]
+
+    The optimized values for the parameters are now stored in ``param``
+    and the optimization behaviour can be assessed by plotting ``cost_rotosolve``,
+    which include the substeps of the Rotosolve optimization.
     The ``full_output`` feature is available for both, ``step`` and ``step_and_cost``.
 
-    The most general form ``RotosolveOptimizer`` is designed to tackle currently is any
-    trigonometric cost function with integer frequencies up to the given value
-    of ``num_freqs`` per parameter. Not all of the integers up to ``num_freqs`` have to
-    be present in the frequency spectrum. In order to tackle equidistant but non-integer
-    frequencies, we recommend rescaling the argument of the function of interest.
+    In general, the frequencies in a QNode will not be integer-valued, requiring us
+    to provide the ``RotosolveOptimizer`` not only with the number of frequencies
+    but their concrete values. For the example QNode above, this happens if the
+    weights are no longer one:
+
+    >>> rot_weights = np.array([0.4, 0.8, 1.2], requires_grad=False)
+    >>> crot_weights = np.array([0.5, 1.0, 1.5], requires_grad=False)
+    >>> spectrum_fn = qml.fourier.qnode_spectrum(cost_function)
+    >>> spectra = spectrum_fn(*param, rot_weights=rot_weights, crot_weights=crot_weights)
+    >>> spectra["rot_param"]
+    {(0,): [-0.4, 0.0, 0.4], (1,): [-0.8, 0.0, 0.8], (2,): [-1.2, 0.0, 1.2]}
+    >>> spectra["crot_param"]
+    {(0,): [-0.5, -0.25, 0.0, 0.25, 0.5], (1,): [-1.0, -0.5, 0.0, 0.5, 1.0], (2,): [-1.5, -0.75, 0.0, 0.75, 1.5]}
+
+    We may provide these spectra instead of ``nums_frequency`` to Rotosolve to
+    enable the optimization of the QNode at these weights:
+
+    >>> param = init_param
+    >>> for step in range(num_steps):
+    ...     param, cost, sub_cost = opt.step_and_cost(
+    ...         cost_function,
+    ...         *param,
+    ...         spectra=spectra,
+    ...         full_output=True,
+    ...         rot_weights = rot_weights,
+    ...         crot_weights = crot_weights,
+    ...     )
+    ...     print(f"Cost before step: {cost}")
+    ...     print(f"Minimization substeps: {np.round(sub_cost, 6)}")
+    Cost before step: 0.09299359486191039
+    Minimization substeps: [-0.268008 -0.713209 -0.24993  -0.871989 -0.907672 -0.907892 -0.940474]
+    Cost before step: -0.9404742138557066
+    Minimization substeps: [-0.940474 -1.       -1.       -1.       -1.       -1.       -1.      ]
+    Cost before step: -1.0
+    Minimization substeps: [-1. -1. -1. -1. -1. -1. -1.]
+
+    As we can see, while the optimization got a bit harder and the optimizer takes a bit longer
+    to converge than previously, Rotosolve was able to adapt to the more complicated
+    dependence on the input arguments and still found the global minimum successfully.
```

Comment:
what a nice and thorough example!

Reply:
@dwierichs you have gotten very good at writing docstrings!

Reply:
Thank you! :)